File: SparkScalaCourse/src/main/scala/com/leo/spark/SpentByCostumer.scala
package com.leo.spark

import org.apache.log4j._
import org.apache.spark._

object SpentByCostumer {

  /** Function to separate lines to a tuple (customerID, value spent) */
  def parseLines(lines: String): (Int, Float) ={
    val lineData = lines.split(",")
    val customer = lineData(0).toInt
    val value = lineData(2).toFloat
    (customer, value)
  }

  def main(args: Array[String]): Unit = {

    val DATA_PATH = "../data/"

     // Remove messages that are not error
    Logger.getLogger("org").setLevel(Level.ERROR)

     // The function
    val totalCostumer = new SparkContext("local[*]", "SpentByCostumer")
      .textFile(DATA_PATH + "customer-orders.csv")
      .map(parseLines)
      .reduceByKey((x, y) => (x + y))
      .map(x => (x._2, x._1))
      .sortByKey()
    totalCostumer.collect().foreach(println)
     //countWordsOccurrence.collect().foreach(println)

     // Creates new spark context FriendsByAge locally
    //val sc = new SparkContext("local[*]", "TotalCostumer")

     // Read data
    //val readFile = sc.textFile(DATA_PATH + "customer-orders.csv")

     // Get information as tuple (customerID, spent)
    //val tupleID_Value = readFile.map(parseLines)

    // Sum the amount in a key
    //val totalByCustomer = tupleID_Value.reduceByKey((x, y) => (x + y))

    // Swap ID and value
    //val swappedData = totalByCustomer.map(x => (x._2, x._1))

    // Sort by key
    //val orderedTotal = swappedData.sortByKey()

   }
}

File: SparkPythonCourse/spent-by-costumer.py
# -*- coding: utf-8 -*-
"""
Created on Sun Dec 26 19:51:48 2021

@author: LeonardoDiasdaRosa
"""

from pyspark import SparkConf, SparkContext

## Data
DATA_PATH = '../data/'
FILE = "customer-orders.csv"

data = DATA_PATH + FILE

## Function to parse the data
def parseData(csvData):
    clientData = csvData.split(",")
    clientID = clientData[0]
    spent = float(clientData[2])
    return (clientID, spent)
    

## Create Spark context
conf = SparkConf().setMaster("local").setAppName("SpentByCostumer")
sc = SparkContext(conf = conf)


## The function
spentByCostumerOrderedByTotal = sc.textFile(data).map(parseData).reduceByKey(lambda x, y: (x + y))\
    .map(lambda x: (x[1], x[0])).sortByKey()

## Read file
# spentData = sc.textFile(data)

## Relevant tuple
# spentPerID_tuple = spentData.map(parseData)

## reduce by key
# spentByClient = spentPerID_tuple.reduceByKey(lambda x, y: (x + y))

## reorder tuple to sort by total spent
# reOrder = spentByClient.map(lambda x: (x[1], x[0]))

## Sort
# spentByCostumerOrderedByTotal = reOrder.sortByKey()

for i in spentByCostumerOrderedByTotal.collect():
    print(i[1], round(i[0], 2))

